In [2]:
import numpy as np
import pandas as pd
import ast

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
credits = pd.read_csv('/content/drive/MyDrive/credits.csv')
keywords = pd.read_csv('/content/drive/MyDrive/keywords.csv')
movies_data = pd.read_csv('/content/drive/MyDrive/movies_metadata.csv')

<ipython-input-6-5f052fb1cf3d>:3: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies_data = pd.read_csv('/content/drive/MyDrive/movies_metadata.csv')


# New Section

In [ ]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [ ]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [ ]:
credits.shape

(45476, 3)

In [ ]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [ ]:
keywords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        46419 non-null  int64 
 1   keywords  46419 non-null  object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


In [ ]:
keywords.shape

(46419, 2)

In [ ]:
movies_data.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [ ]:
movies_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

- Belongs_to_collection(40972), homepage(37684), tagline(25054)
- imdb_id(17), original_language(11), popularity(5), production_companies(3), production_contries(3), release_date(87),revenue(6), spoken_languages(6), status(87), title(6), video(6), vote_average(6), vote_count(6)
- overview(954), poster_path(381), runtime(269),

= Most of them contain null values but majority features have less than 100 points of null values. 3 have very high null values and 3 have null values between 100-1000

= Now, How should I proceed. "Should I first remove null values or irrelevent features"

In [ ]:
print(type(movies_data))
print(type(keywords))
print(type(credits))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


In [ ]:
#needed information -
#id
#title
#overview
#genres
#keywords
#cast
#crew

In [ ]:
credits['id'].values

array([   862,   8844,  15602, ...,  67758, 227506, 461257])

In [ ]:
credits.head()['id'].values

array([  862,  8844, 15602, 31357, 11862])

Q = Are all three data similar? I mean id of credits data matching with title of movies_data

In [ ]:
movies = movies_data[['id','title','overview', 'genres']]

In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                              genres  
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  
4                     [{'id': 35, 'name': 'Comedy'}]

In [ ]:
type(movies)

pandas.core.frame.DataFrame

In [ ]:
movies.reset_index(drop=True)

id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
45461  439050                       Subdue   
45462  111109          Century of Birthing   
45463   67758                     Betrayal   
45464  227506             Satan Triumphant   
45465  461257                     Queerama   

                                                overview  \
0      Led by Woody, Andy's toys live happily in his ...   
1      When siblings Judy and Peter discover an encha...   
2      A family wedding reignites the ancient feud be...   
3      Cheated on, mistreated and stepped on, the wom...   
4      Just when George Banks has recovered from his ...   
...                                                  ...   
45461        Rising and falling between a man and woman.   
45462  An artist struggles to finish his work while a...   
45463  When one of her hits goes wrong, a professiona...   
45464  In a small town live two brothers, one a minis...   
45465  50 years after decriminalisation of homosexual...   

                                                  genres  
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  
4                         [{'id': 35, 'name': 'Comedy'}]  
...                                                  ...  
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...  
45462                      [{'id': 18, 'name': 'Drama'}]  
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...  
45464                                                 []  
45465                                                 []  

[45466 rows x 4 columns]

In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                                              genres  
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  
4                     [{'id': 35, 'name': 'Comedy'}]

- You extracted name in the genre column but what about the id. Is it not the same as id column in the movie data or Are both id's different

In a movie dataset, the presence of 'id' in two different features, such as the main 'id' feature and within the 'genres' feature as part of a dictionary, generally indicates that they serve different purposes. both instances use the term 'id,' they likely refer to different entities: one for uniquely identifying movies and the other for uniquely identifying genres associated with those movies.

In [ ]:
movies.isnull().sum()

id            0
title         6
overview    954
genres        0
dtype: int64

In [ ]:
movies.dropna(inplace=True)

<ipython-input-10-4d2d825739df>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.dropna(inplace=True)


In [ ]:
movies.duplicated().sum()

30

In [ ]:
movies.iloc[0].genres

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [ ]:
def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [ ]:
movies['genres'] = movies['genres'].apply(convert)

<ipython-input-14-fd9a96deee9b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['genres'] = movies['genres'].apply(convert)


In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
movies.iloc[:,2:3]

overview
0      Led by Woody, Andy's toys live happily in his ...
1      When siblings Judy and Peter discover an encha...
2      A family wedding reignites the ancient feud be...
3      Cheated on, mistreated and stepped on, the wom...
4      Just when George Banks has recovered from his ...
...                                                  ...
45461        Rising and falling between a man and woman.
45462  An artist struggles to finish his work while a...
45463  When one of her hits goes wrong, a professiona...
45464  In a small town live two brothers, one a minis...
45465  50 years after decriminalisation of homosexual...

[44506 rows x 1 columns]

In [ ]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [ ]:
keywords['keywords'] = keywords['keywords'].apply(convert)

In [ ]:
keywords.head()

id                                           keywords
0    862  [jealousy, toy, boy, friendship, friends, riva...
1   8844  [board game, disappearance, based on children'...
2  15602  [fishing, best friend, duringcreditsstinger, o...
3  31357  [based on novel, interracial relationship, sin...
4  11862  [baby, midlife crisis, confidence, aging, daug...

Q = How did you get keywords data? It's looking like id for keywords and movies data is same. Are they extracted from the overview or we took a bunch of keywords by ourselves and later we will use them to find find similar keywords in overview or any other feature to train our model.

In [ ]:
def convert3(obj):
    L = []
    counter = 0
    for i in ast.literal_eval(obj):
        if counter != 3:   #why??
            L.append(i['name'])
            counter += 1
        else:
            break
    return L

In [ ]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [ ]:
credits['cast'] = credits['cast'].apply(convert3)

In [ ]:
credits.head()

cast  \
0                [Tom Hanks, Tim Allen, Don Rickles]   
1     [Robin Williams, Jonathan Hyde, Kirsten Dunst]   
2         [Walter Matthau, Jack Lemmon, Ann-Margret]   
3  [Whitney Houston, Angela Bassett, Loretta Devine]   
4         [Steve Martin, Diane Keaton, Martin Short]   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [ ]:
def fetch_director(obj):
    L = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L

In [ ]:
credits['crew'] = credits['crew'].apply(fetch_director)

In [ ]:
credits['crew'].head()

0      [John Lasseter]
1       [Joe Johnston]
2      [Howard Deutch]
3    [Forest Whitaker]
4      [Charles Shyer]
Name: crew, dtype: object

In [ ]:
credits.head()

cast               crew     id
0                [Tom Hanks, Tim Allen, Don Rickles]    [John Lasseter]    862
1     [Robin Williams, Jonathan Hyde, Kirsten Dunst]     [Joe Johnston]   8844
2         [Walter Matthau, Jack Lemmon, Ann-Margret]    [Howard Deutch]  15602
3  [Whitney Houston, Angela Bassett, Loretta Devine]  [Forest Whitaker]  31357
4         [Steve Martin, Diane Keaton, Martin Short]    [Charles Shyer]  11862

In [ ]:
keywords.head()

id                                           keywords
0    862  [jealousy, toy, boy, friendship, friends, riva...
1   8844  [board game, disappearance, based on children'...
2  15602  [fishing, best friend, duringcreditsstinger, o...
3  31357  [based on novel, interracial relationship, sin...
4  11862  [baby, midlife crisis, confidence, aging, daug...

In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
movies['overview'] = movies['overview'].apply(lambda x : x.split())

<ipython-input-21-6599156e7507>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['overview'] = movies['overview'].apply(lambda x : x.split())


In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
movies['genres'] = movies['genres'].apply(lambda x: [i.replace(" ","") for i in x])
keywords['keywords'] = keywords['keywords'].apply(lambda x: [i.replace(" ","") for i in x])
credits['cast'] = credits['cast'].apply(lambda x: [i.replace(" ","") for i in x])
credits['crew'] = credits['crew'].apply(lambda x: [i.replace(" ","") for i in x])

<ipython-input-22-16bca8c66636>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['genres'] = movies['genres'].apply(lambda x: [i.replace(" ","") for i in x])


In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
credits.head()

cast              crew     id
0                [TomHanks, TimAllen, DonRickles]    [JohnLasseter]    862
1     [RobinWilliams, JonathanHyde, KirstenDunst]     [JoeJohnston]   8844
2        [WalterMatthau, JackLemmon, Ann-Margret]    [HowardDeutch]  15602
3  [WhitneyHouston, AngelaBassett, LorettaDevine]  [ForestWhitaker]  31357
4         [SteveMartin, DianeKeaton, MartinShort]    [CharlesShyer]  11862

In [ ]:
keywords.head()

id                                           keywords
0    862  [jealousy, toy, boy, friendship, friends, riva...
1   8844  [boardgame, disappearance, basedonchildren'sbo...
2  15602  [fishing, bestfriend, duringcreditsstinger, ol...
3  31357  [basedonnovel, interracialrelationship, single...
4  11862  [baby, midlifecrisis, confidence, aging, daugh...

In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
movies_id = np.array(movies['id'])
len(movies_id)

44506

In [ ]:
movies_id

array(['862', '8844', '15602', ..., '67758', '227506', '461257'],
      dtype=object)

In [ ]:
credits_id = np.array(credits['id'])
print(len(credits_id))
credits_id

45476


array([   862,   8844,  15602, ...,  67758, 227506, 461257])

In [ ]:
keywords_id = np.array(keywords['id'])
print(len(keywords_id))
keywords_id

46419


array([   862,   8844,  15602, ...,  67758, 227506, 461257])

In [ ]:
movies_id = np.array([int(item) for item in movies_id])
credits_id = np.array([int(item) for item in credits_id])
keywords_id = np.array([int(item) for item in keywords_id])  #checking all the id's are integer or not

In [ ]:
common1 = np.intersect1d(movies_id,credits_id)
print(len(common1))
common1

44475


array([     2,      3,      5, ..., 467731, 468343, 469172])

In [ ]:
common_id = np.intersect1d(common1,keywords_id)
print(len(common_id))
common_id

44475


array([     2,      3,      5, ..., 467731, 468343, 469172])

In [ ]:
new_df = pd.DataFrame(common_id, columns = ['id'])
new_df

id
0           2
1           3
2           5
3           6
4          11
...       ...
44470  464819
44471  465044
44472  467731
44473  468343
44474  469172

[44475 rows x 1 columns]

In [ ]:
len(movies.id)

44506

In [ ]:
credits.head()

cast              crew     id
0                [TomHanks, TimAllen, DonRickles]    [JohnLasseter]    862
1     [RobinWilliams, JonathanHyde, KirstenDunst]     [JoeJohnston]   8844
2        [WalterMatthau, JackLemmon, Ann-Margret]    [HowardDeutch]  15602
3  [WhitneyHouston, AngelaBassett, LorettaDevine]  [ForestWhitaker]  31357
4         [SteveMartin, DianeKeaton, MartinShort]    [CharlesShyer]  11862

In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
# movies = movies.reset_index(drop = True)
# data_types = movies['id'].apply(type)
# data_types.value_counts()

In [ ]:
#I tried to create a completely new dataframe called new_df with one column having the common id's of all 3 dataframes and
#other column 'title' having 'title' of the movie corresponding to that id and 'tag' having sum of 'overview', 'cast', 'crew'
# and 'keywords'

#it is taking a LOT of time to process. I also tried different ways to remove NaN errors but no progress :(
#tried using merge, pd.concat, pd.merge, converted int64 to object and tried etc. None worked. If you guys can do that then
#preprocessing can be much easier without a need to define arrays for identities and stuff. Lets continue tomorrow! Good night.

In [ ]:
import pandas as pd

# Assuming 'df' is your DataFrame and 'column_name' is the column you want to check
column_name = 'id'  # Replace with the actual column name

# Convert the column to numeric, coercing errors to NaN for non-numeric values
numeric_values = pd.to_numeric(movies['id'], errors='coerce')

# Count the number of numeric and non-numeric values
numeric_count = numeric_values.notna().sum()
non_numeric_count = numeric_values.isna().sum()

print(f"Numeric values: {numeric_count}")
print(f"Non-numeric values: {non_numeric_count}")

Numeric values: 44506
Non-numeric values: 0


In [ ]:
keywords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        46419 non-null  int64 
 1   keywords  46419 non-null  object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


In [ ]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [ ]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44506 entries, 0 to 45465
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        44506 non-null  object
 1   title     44506 non-null  object
 2   overview  44506 non-null  object
 3   genres    44506 non-null  object
dtypes: object(4)
memory usage: 1.7+ MB


In [ ]:
movies['id'] = movies['id'].astype(int)

<ipython-input-31-5ef8ba7aed37>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['id'] = movies['id'].astype(int)


In [ ]:
movies['id'].dtypes

dtype('int64')

In [ ]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44506 entries, 0 to 45465
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        44506 non-null  int64 
 1   title     44506 non-null  object
 2   overview  44506 non-null  object
 3   genres    44506 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.7+ MB


In [ ]:
movies.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  
0   [Animation, Comedy, Family]  
1  [Adventure, Fantasy, Family]  
2             [Romance, Comedy]  
3      [Comedy, Drama, Romance]  
4                      [Comedy]

In [ ]:
df0 = pd.merge(movies, credits, on = 'id', how = 'inner')
df = pd.merge(df0, keywords, on = 'id', how = 'inner')
df.iloc[0:1000,:]

id                                   title  \
0      862                               Toy Story   
1     8844                                 Jumanji   
2    15602                        Grumpier Old Men   
3    31357                       Waiting to Exhale   
4    11862             Father of the Bride Part II   
..     ...                                     ...   
995   6878  Homeward Bound: The Incredible Journey   
996  15944                          The Shaggy Dog   
997  18444                   Swiss Family Robinson   
998  20723                          That Darn Cat!   
999    173            20,000 Leagues Under the Sea   

                                              overview  \
0    [Led, by, Woody,, Andy's, toys, live, happily,...   
1    [When, siblings, Judy, and, Peter, discover, a...   
2    [A, family, wedding, reignites, the, ancient, ...   
3    [Cheated, on,, mistreated, and, stepped, on,, ...   
4    [Just, when, George, Banks, has, recovered, fr...   
..                                                 ...   
995  [Remake, of, the, popular, Disney, classic,, t...   
996  [Through, an, ancient, spell,, a, boy, changes...   
997  [After, being, shipwrecked,, the, Robinson, fa...   
998  [A, young, woman, suspects, foul, play, when, ...   
999  [A, ship, sent, to, investigate, a, wave, of, ...   

                                 genres  \
0           [Animation, Comedy, Family]   
1          [Adventure, Fantasy, Family]   
2                     [Romance, Comedy]   
3              [Comedy, Drama, Romance]   
4                              [Comedy]   
..                                  ...   
995  [Adventure, Comedy, Drama, Family]   
996                    [Comedy, Family]   
997                 [Adventure, Family]   
998             [Drama, Family, Comedy]   
999  [Adventure, Drama, ScienceFiction]   

                                               cast                crew  \
0                  [TomHanks, TimAllen, DonRickles]      [JohnLasseter]   
1       [RobinWilliams, JonathanHyde, KirstenDunst]       [JoeJohnston]   
2          [WalterMatthau, JackLemmon, Ann-Margret]      [HowardDeutch]   
3    [WhitneyHouston, AngelaBassett, LorettaDevine]    [ForestWhitaker]   
4           [SteveMartin, DianeKeaton, MartinShort]      [CharlesShyer]   
..                                              ...                 ...   
995           [MichaelJ.Fox, SallyField, DonAmeche]     [DuwayneDunham]   
996           [FredMacMurray, JeanHagen, TommyKirk]     [CharlesBarton]   
997     [JohnMills, DorothyMcGuire, JamesMacArthur]        [KenAnnakin]   
998        [HayleyMills, DeanJones, DorothyProvine]   [RobertStevenson]   
999            [KirkDouglas, JamesMason, PaulLukas]  [RichardFleischer]   

                                              keywords  
0    [jealousy, toy, boy, friendship, friends, riva...  
1    [boardgame, disappearance, basedonchildren'sbo...  
2    [fishing, bestfriend, duringcreditsstinger, ol...  
3    [basedonnovel, interracialrelationship, single...  
4    [baby, midlifecrisis, confidence, aging, daugh...  
..                                                 ...  
995  [basedonnovel, cat, friendship, remake, dog, j...  
996                             [magic, dog, sheepdog]  
997  [treehouse, island, shipwreck, pirategang, swi...  
998  [cat, hostage, kidnapping, map, sister, suburb...  
999  [diving, ocean, submarine, julesverne, captain...  

[1000 rows x 7 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45629 entries, 0 to 45628
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        45629 non-null  int64 
 1   title     45629 non-null  object
 2   overview  45629 non-null  object
 3   genres    45629 non-null  object
 4   cast      45629 non-null  object
 5   crew      45629 non-null  object
 6   keywords  45629 non-null  object
dtypes: int64(1), object(6)
memory usage: 2.8+ MB


- 45629 entries and in this dataset there is no null values How?? 45629
- Initially credits(45476), keywords(46419), movie_data(45466)
- common id are 44475 which mean that there are some uncommon id's which the new df contains but what about the values of others which doenn't contain that id values What is that value? if there is no value then what is filled there.
- Should I proceed with this data?

In [ ]:
#df['tags'] = df['overview'] + df['genres'] + df['cast'] + df['crew'] + df['keywords']

In [ ]:
#new_df = df[['id','title','tags']]

In [ ]:
#new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))

In [ ]:

#new_df.head()

In [ ]:
df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  \
0   [Animation, Comedy, Family]   
1  [Adventure, Fantasy, Family]   
2             [Romance, Comedy]   
3      [Comedy, Drama, Romance]   
4                      [Comedy]   

                                             cast              crew  \
0                [TomHanks, TimAllen, DonRickles]    [JohnLasseter]   
1     [RobinWilliams, JonathanHyde, KirstenDunst]     [JoeJohnston]   
2        [WalterMatthau, JackLemmon, Ann-Margret]    [HowardDeutch]   
3  [WhitneyHouston, AngelaBassett, LorettaDevine]  [ForestWhitaker]   
4         [SteveMartin, DianeKeaton, MartinShort]    [CharlesShyer]   

                                            keywords  
0  [jealousy, toy, boy, friendship, friends, riva...  
1  [boardgame, disappearance, basedonchildren'sbo...  
2  [fishing, bestfriend, duringcreditsstinger, ol...  
3  [basedonnovel, interracialrelationship, single...  
4  [baby, midlifecrisis, confidence, aging, daugh...

In [ ]:
import pandas as pd

# combining crew, genres, keywords into single feature
#df['description'] = df.apply(lambda row: row['crew'] + row['genres'] + row['keywords'], axis=1)
df['description'] = df.apply(lambda row: row['genres'] + row['keywords'], axis=1)

df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                            overview  \
0  [Led, by, Woody,, Andy's, toys, live, happily,...   
1  [When, siblings, Judy, and, Peter, discover, a...   
2  [A, family, wedding, reignites, the, ancient, ...   
3  [Cheated, on,, mistreated, and, stepped, on,, ...   
4  [Just, when, George, Banks, has, recovered, fr...   

                         genres  \
0   [Animation, Comedy, Family]   
1  [Adventure, Fantasy, Family]   
2             [Romance, Comedy]   
3      [Comedy, Drama, Romance]   
4                      [Comedy]   

                                             cast              crew  \
0                [TomHanks, TimAllen, DonRickles]    [JohnLasseter]   
1     [RobinWilliams, JonathanHyde, KirstenDunst]     [JoeJohnston]   
2        [WalterMatthau, JackLemmon, Ann-Margret]    [HowardDeutch]   
3  [WhitneyHouston, AngelaBassett, LorettaDevine]  [ForestWhitaker]   
4         [SteveMartin, DianeKeaton, MartinShort]    [CharlesShyer]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [boardgame, disappearance, basedonchildren'sbo...   
2  [fishing, bestfriend, duringcreditsstinger, ol...   
3  [basedonnovel, interracialrelationship, single...   
4  [baby, midlifecrisis, confidence, aging, daugh...   

                                         description  
0  [Animation, Comedy, Family, jealousy, toy, boy...  
1  [Adventure, Fantasy, Family, boardgame, disapp...  
2  [Romance, Comedy, fishing, bestfriend, duringc...  
3  [Comedy, Drama, Romance, basedonnovel, interra...  
4  [Comedy, baby, midlifecrisis, confidence, agin...

In [ ]:
df['director'] = df['crew'].apply(lambda x: f"DirectorIs{', '.join(x)}").apply(lambda x: [x])

In [ ]:
df['description'] = df.apply(lambda row: row['description'] + row['director'], axis=1)

In [ ]:
final_df = df[['id', 'title', 'description']]
final_df.head()

id                        title  \
0    862                    Toy Story   
1   8844                      Jumanji   
2  15602             Grumpier Old Men   
3  31357            Waiting to Exhale   
4  11862  Father of the Bride Part II   

                                         description  
0  [Animation, Comedy, Family, jealousy, toy, boy...  
1  [Adventure, Fantasy, Family, boardgame, disapp...  
2  [Romance, Comedy, fishing, bestfriend, duringc...  
3  [Comedy, Drama, Romance, basedonnovel, interra...  
4  [Comedy, baby, midlifecrisis, confidence, agin...

In [ ]:
final_df['embeddings']=np.zeros(len(final_df))

<ipython-input-73-ca9c56a5ee2c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['embeddings']=np.zeros(len(final_df))


In [ ]:
pip install -q transformers tqdm datasets huggingface_hub

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 17.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 493.7/493.7 kB 24.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.1/311.1 kB 25.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 36.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 41.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 13.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 18.6 MB/s eta 0:00:00


In [ ]:
from huggingface_hub import notebook_login
notebook_login()

In [ ]:
import torch
from transformers import AutoTokenizer, AutoModel
import pandas as pd
import numpy as np
from tqdm import tqdm

# Assuming try_df is your DataFrame

# Load the tokenizer and model
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
model = AutoModel.from_pretrained('bert-base-uncased')

# Create a function to get embeddings
def get_embedding_batch(descriptions):
    tokens = tokenizer(descriptions, return_tensors='pt', padding=True, truncation=True)
    with torch.no_grad():
        output = model(**tokens)
    return output.last_hidden_state.mean(dim=1).squeeze().detach().numpy()

# Batch size
batch_size = 2500

# Total number of rows in the DataFrame
start_row=40000
end_row=len(final_df)

# Process the DataFrame in batches
for i in range(start_row, end_row, batch_size):
    # Get a batch of descriptions
    final_df['embeddings'].iloc[i:i+batch_size] = final_df['description'].iloc[i:i+batch_size].progress_apply(get_embedding_batch)

    # Save the embeddings to a Hugging Face model for the batch
    hf_model_batch = {
        'titles': list(final_df['title'].iloc[i:i+batch_size]),
        'embeddings': list(final_df['embeddings'].iloc[i:i+batch_size])
    }

    # Save the batch to a file
    csv_file_name = f'hf_model_batch_{i // batch_size}.csv'
    final_df.to_csv(csv_file_name, index=False)

    # Download the file
    from google.colab import files
    files.download(csv_file_name)

Creating Embeddings: 100%|██████████| 2500/2500 [11:21<00:00,  3.67it/s]
<ipython-input-81-abd913284899>:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df['embeddings'].iloc[i:i+batch_size] = final_df['description'].iloc[i:i+batch_size].progress_apply(get_embedding_batch)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Creating Embeddings: 100%|██████████| 2500/2500 [10:51<00:00,  3.84it/s]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Creating Embeddings: 100%|██████████| 629/629 [02:31<00:00,  4.16it/s]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Now we have made another .ipynb file named ME781_Gradio for better user interface. So to run the recommendation system you only have to run the gradio file and not this file.